src/black/__init__.py

```diff
@@ -749,7 +749,12 @@ def get_sources(
 
         # Compare the logic here to the logic in `gen_python_files`.
         if is_stdin or path.is_file():
-            root_relative_path = get_root_relative_path(path, root, report)
+            # From https://stackoverflow.com/questions/42513056/how-to-get-absolute-path-of-a-pathlib-path-object  # noqa: B950
+            # suggest to use resolve instead of absolute prior to Python 3.11
+            # to get the absolute path.
+            # This change is to fix bug:
+            # https://github.com/psf/black/issues/4209
+            root_relative_path = get_root_relative_path(path.resolve(), root, report)
```

Comment:
Looks like `.resolve()` will raise an error in some cases if the path doesn't exist or is a broken symbolic link. Is that the right behavior here?

I'm less concerned about using `.absolute()` than that StackOverflow answer. If it's documented in recent versions, we can use `.absolute()` here; it's not going to change behavior now in Python 3.8 and 3.9.

Reply:
Maybe this is the right solution; I don't know much about Windows paths so I can't tell. I just want to make sure we use a solution that's unlikely to cause new problems in other contexts.

Reply:
the resolve raise error before python 3.6, which is already deprecate. Need we support Python prior to 3.7 inclusively?

Reply:
We don't care about 3.6 or 3.7.

Reply:
`resolve` hurts performance and resolving symlinks can result in unexpected behaviour. I'd rather e.g. inline the right version of relative_to on 3.8

Reply:
what's your suggestion to fix this bug?

Reply:
just tried  Path("\\\\?\\d:\\git\\Olive\\setup.py").absolute().relative_to(Path("d:\\git\\Olive")) on Python 3.10 and 3.11. 
both of them complains like:
![image](https://github.com/psf/black/assets/817030/7e5b6072-38d0-4d05-8a2d-c6519fd6e99f)

Reply:
Would you mind testing https://github.com/psf/black/pull/4222 and seeing if it works well for your use case?

Reply:
sorry for late response since Chinese New Year vacation. will try and update the thread.

Reply:
PR https://github.com/psf/black/pull/4222 fix the issue. will close this PR.